### PR TITLE
Add code action annotate

### LINKED ISSUE #396: Feature request: add "annotate value with type" code action
I'd like to have a code action that would turn something like `let foo = 1 in` into `let (foo: int) = 1 in` when called on `foo`.

### PR DESCRIPTION
**Pull request by Khady**
New code action `annotate`, that adds a type next to the value on which the action is called. The goal is to be able to turn `let x = x + 1` into `let (x : int) = x + 1`.

I'm missing a few things before the PR is ready.

1. [x] I don't know how to get in the file what is in the original range of text
1. [x] I need to get the name of the type only and not its detail, for example in this snippet I should get `let f (x : t) = (x : t)` and not `let f (x : t) = (x : type t = Foo of int | Bar of bool)`
  ```
type t = Foo of int | Bar of bool

let f (x : t) = x
  ```

fix #396

### CODE REVIEW

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts [hunk 1]
```diff
@@ -66,6 +66,30 @@ let f (x : t) = x
         kind: "destruct",
         title: "Destruct",
       },
+      {
+        edit: {
+          changes: {
+            "file:///test.ml": [
+              {
+                newText: "(x : type t = Foo of int | Bar of bool)",
```
Comment (Khady) [inline ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts:74]: here is the place where the type is not displayed as it should

Comment (rgrinberg) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts:74]: Yes, you should only enable this code actions on expressions.

Comment (ulugbekna) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts:74]: Would querying the type of an expression using mconfig verbosity = 0 (see `ocaml_lsp_server/src/document.ml:136`) help here?

Then we go from 

```ocaml
let (not_supported : ('a, Jsonrpc.Response.Error.t) result

type ('a, 'b) result = Ok of 'a | Error of 'b) =
  Error
    (make_error ~code:InternalError ~message:"Request not supported yet!" ())
```

to correct

```ocaml
let (not_supported : ('a, Jsonrpc.Response.Error.t) result) =
  Error
    (make_error ~code:InternalError ~message:"Request not supported yet!" ())
```

Comment (Khady) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts]: Thanks, I found a way to reduce the verbosity and it seems to help

File: ocaml-lsp-server/src/code_actions/action_annotate.ml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+let code_action doc (params : CodeActionParams.t) =
+  let uri = Uri.t_of_yojson (`String params.textDocument.uri) in
+  let command =
+    let start = Position.logical params.range.start in
+    Query_protocol.Type_enclosing (None, start, None)
+  in
+  let open Fiber.O in
+  let+ res =
+    Document.with_pipeline doc (fun pipeline ->
+        let config = Mpipeline.final_config pipeline in
```
Comment (Khady) [inline ocaml-lsp-server/src/code_actions/action_annotate.ml:36]: I don't know if I am using the right function to get the config

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts [hunk 3]
```diff
@@ -143,4 +167,169 @@ let f (x : t) = x
...
+type t = Foo of int | Bar of bool
+
+let f x = Foo x
+`,
+      "file:///test.ml",
+    );
+    let start = Types.Position.create(2, 6);
+    let end = Types.Position.create(2, 7);
+    let actions = await codeAction("file:///test.ml", start, end);
+    expect(actions).toMatchObject([
```
Comment (rgrinberg) [inline ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts:183]: Can you filter this list to only show the code action you're interested in?

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts [hunk 4]
```diff
@@ -143,4 +167,169 @@ let f (x : t) = x
...
+                    character: 15,
+                    line: 3,
+                  },
+                },
+              },
+            ],
+          },
+        },
+        kind: "destruct",
+        title: "Destruct",
```
Comment (rgrinberg) [inline ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts:307]: Same here. We don't need to see destruct here.

**Comment by rgrinberg** [comment]
What happens when you call this code action on this snippet (where `[ ]` is the cursor)?

```ocaml
type x =
   | Foo of int
   | [B]az of string
```

in other words, you try to execute the code action in a non expression context?

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts (hunk 3 above)
Comment (Khady) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-codeAction.test.ts]: Solved. The diff displayed is less actionable (everything is on a single line), but it's not dramatic

**Comment by Khady** [comment]
I've added a test to show the behavior in a non expression context. For now it would change `Baz` into `(Baz : string -> x)`. I don't know how to check the context to forbid the annotation in this case.

**Comment by rgrinberg** [comment]
>  I don't know how to check the context to forbid the annotation in this case.

You can refer to how destruct works. We make sure that it works only expressions. This might require some changes to our merlin fork, but that's fine.

**Comment by Khady** [comment]
I haven't found where to do that. Could you give me a pointer?

While digging I found this [`type node`](https://github.com/rgrinberg/merlin/blob/f6954e953b4168e6a798a0255d6a2dfbd868a3c6/src/ocaml/merlin_specific/browse_raw.mli#L49-L98) which I believe is pretty close to what we would want to filter on. But didn't completely figured how to get it yet.

**Comment by rgrinberg** [comment]
You can obtain it with:

```
    let browse = Mbrowse.of_typedtree (Mtyper.get_typedtree typer) in
    let nodes = Mbrowse.enclosing pos_start [browse] in
```

Then inspect the first node to see if its an expression.

**Comment by Khady** [comment]
Thanks. I think that I got it working.

File: ocaml-lsp-server/src/code_actions/action_annotate.ml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        `Invalid)
+
+let get_source_text doc (loc : Loc.t) =
+  let open Option.O in
+  let source = Document.source doc in
+  let* start = Position.of_lexical_position loc.loc_start in
+  let+ end_ = Position.of_lexical_position loc.loc_end in
+  let (`Offset start) = Msource.get_offset source (Position.logical start) in
+  let (`Offset end_) = Msource.get_offset source (Position.logical end_) in
+  String.sub (Msource.text source) ~pos:start ~len:(end_ - start)
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/code_actions/action_annotate.ml:23]: You can use `String.drop` here.

File: ocaml-lsp-server/src/code_actions/action_annotate.ml [hunk 6]
```diff
@@ -0,0 +1,63 @@
+open Import
+
+let action_kind = "annotate"
+
+let check_context doc pos_start =
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/code_actions/action_annotate.ml:5]: I'd choose a name that's closer to the property being checked. `check_typeable_context` or something.

File: ocaml-lsp-server/src/code_actions/action_annotate.ml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  let uri = Uri.t_of_yojson (`String params.textDocument.uri) in
+  let pos_start = Position.logical params.range.start in
+  let* context = check_context doc pos_start in
+  match context with
+  | Error e -> Fiber.return (Error (Jsonrpc.Response.Error.of_exn e))
+  | Ok `Invalid -> Fiber.return (Ok None)
+  | Ok `Valid -> (
+    let command = Query_protocol.Type_enclosing (None, pos_start, None) in
+    let+ res =
+      Document.with_pipeline doc (fun pipeline ->
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/code_actions/action_annotate.ml:49]: The execution of this code action makes two calls to `with_pipeline`. That is slightly inefficient, but there's a bit of a bigger problem - when you call `with_pipeline`, you're yielding until the merlin thread is ready. So it's possible that there might be other requests served in the meantime that will invalidate your check for expression context. It should be possible to refactor the code to only call `with_pipeline` once.

File: ocaml-lsp-server/src/code_actions/action_annotate.ml (hunk 5 above)
Comment (Khady) [reply ocaml-lsp-server/src/code_actions/action_annotate.ml:23]: I don't think so, drop is only dropping the beginning of the string while I want to drop at the beginning and at the end?

Comment (rgrinberg) [reply ocaml-lsp-server/src/code_actions/action_annotate.ml:23]: Nvm, you're right.

**Comment by rgrinberg** [comment]
Oh, and we need a CHANGES entry for this.

**Comment by Khady** [comment]
Thanks for being so reactive. I moved to a single with_pipeline and added a line to CHANGES.

**Comment by rgrinberg** [comment]
Thanks for your contribution.

**Pull request merged**